doc/rpmspec.8
.TH "RPMSPEC" "8" "29 October 2010" "Red Hat, Inc"
.SH NAME
rpmspec \- RPM Spec Tool
.SH SYNOPSIS
.SS "QUERYING SPEC FILES:"
.PP

\fBrpmspec\fR {\fB-q|--query\fR} [\fBselect-options\fR] [\fBquery-options\fR] \fB\fISPEC_FILE\fB\fR\fI ...\fR

.SS "PARSING SPEC FILES TO STDOUT:"
.PP

\fBrpmspec\fR {\fB-P|--parse\fR} \fB\fISPEC_FILE\fB\fR\fI ...\fR

\fBrpmspec\fR {\fB--shell\fR}\fR

.SH DESCRIPTION
.PP
\fBrpmspec\fR is a tool for querying a spec file. More specifically for querying hypothetical packages which would be created from the given spec file. So querying a spec file with \fBrpmspec\fR is similar to querying a package built from that spec file. But is is not identical. With \fBrpmspec\fR you can't query all fields which you can query from a built package. E. g. you can't query BUILDTIME with \fBrpmspec\fR for obvious reasons. You also cannot query other fields automatically generated during a build of a package like auto generated dependencies.

.SS "select-options"
.PP

 [\fB\--rpms\fB\fR]
 [\fB\--srpm\fB\fR]

.SS "query-options"
.PP

 [\fB--qf,--queryformat \fIQUERYFMT\fB\fR]
 [\fB--target \fITARGET_PLATFORM\fB\fR]
 
.SS "QUERY OPTIONS"
.PP
The general form of an rpm spec query command is 
.PP

\fBrpm\fR {\fB-q|--query\fR} [\fBselect-options\fR] [\fBquery-options\fR]

.PP
You may specify the format that the information should be
printed in. To do this, you use the

 \fB--qf|--queryformat\fR \fB\fIQUERYFMT\fB\fR

option, followed by the \fIQUERYFMT\fR format string.
See \fBrpm(8)\fR for details.
.PP

.SS "SELECT OPTIONS"
.PP
 \fB--rpms\fR
Operate on the all binary package headers generated from spec.
 \fB--builtrpms\fR
Operate only on the binary package headers of packages which would be built from spec. That means ignoring package headers of packages that won't be built from spec i. e. ignoring package headers of packages without file section.
 \fB--srpm\fR
Operate on the source package header(s) generated from spec.

.SH EXAMPLES
.PP
Get list of binary packages which would be generated from the rpm spec file:
.PP
.RS 4
.nf
 $ rpmspec -q rpm.spec
 rpm-4.11.3-3.fc20.x86_64
 rpm-libs-4.11.3-3.fc20.x86_64
 rpm-build-libs-4.11.3-3.fc20.x86_64
 ...
.RE
.PP
Get summary infos for single binary packages generated from the rpm spec file:
.PP
.RS 4
.nf
 $ rpmspec -q --qf "%{name}: %{summary}\\n" rpm.spec
 rpm: The RPM package management system
 rpm-libs: Libraries for manipulating RPM packages
 rpm-build-libs: Libraries for building and signing RPM packages
 ...
.RE
.PP
Get the source package which would be generated from the rpm spec file:
.PP
.RS 4
.nf
 $ rpmspec -q --srpm rpm.spec
 rpm-4.11.3-3.fc20.x86_64
.RE
.PP
Parse the rpm spec file to stdout:
.PP
.RS 4
.nf
 $ rpmspec -P rpm.spec
 Summary: The RPM package management system
 Name: rpm
 Version: 4.14.0
 ...
.RE
.PP
Run interactive macro shell for debugging macros
.RS 4
.nf
 $ rpmspec --shell
 > %define foo bar
 > %foo
 bar
 > %(date)
 Tue Apr 13 03:55:37 PM EEST 2021
 >
.RE
.PP
Run interactive macro shell in spec context
.RS 4
.nf
 $ rpmspec --shell popt.spec
> %name
popt
> %version
1.18
>
.RE
.SH "SEE ALSO"
.nf
\fBpopt\fR(3),
\fBrpm\fR(8),
\fBrpmdb\fR(8),
\fBrpmkeys\fR(8),
\fBrpmsign\fR(8),
\fBrpm2cpio\fR(8),
\fBrpmbuild\fR(8),
.fi

\fBrpmspec --help\fR - as rpm supports customizing the options via popt aliases 
it's impossible to guarantee that what's described in the manual matches 
what's available.


\fBhttp://www.rpm.org/ <URL:http://www.rpm.org/>
\fR
.SH "AUTHORS"

.nf
Marc Ewing <marc@redhat.com>
Jeff Johnson <jbj@redhat.com>
Erik Troan <ewt@redhat.com>
Panu Matilainen <pmatilai@redhat.com>
.fi


rpmspec.c
#include "system.h"

#include <rpm/rpmcli.h>
#include <rpm/rpmbuild.h>
#include <rpm/rpmlog.h>
#include <rpm/rpmts.h>

#ifdef HAVE_READLINE
#include <readline/readline.h>
#include <readline/history.h>
#endif

#include "cliutils.h"

#include "debug.h"

enum modes {
    MODE_UNKNOWN	= 0,
    MODE_QUERY		= (1 <<  0),
    MODE_PARSE		= (1 <<  1),
    MODE_SHELL		= (1 <<  2),
};

static int mode = MODE_UNKNOWN;
static int source = RPMQV_SPECRPMS;
char *queryformat = NULL;

static struct poptOption specOptsTable[] = {
    { "parse", 'P', POPT_ARG_VAL, &mode, MODE_PARSE,
	N_("parse spec file(s) to stdout"), NULL },
    { "query", 'q', POPT_ARG_VAL, &mode, MODE_QUERY,
	N_("query spec file(s)"), NULL },
#if HAVE_READLINE
    { "shell", 0, POPT_ARG_VAL, &mode, MODE_SHELL,
	N_("interactive macro shell"), NULL },
#endif
    { "rpms", 0, POPT_ARG_VAL, &source, RPMQV_SPECRPMS,
	N_("operate on binary rpms generated by spec (default)"), NULL },
    { "builtrpms", 0, POPT_ARG_VAL, &source, RPMQV_SPECBUILTRPMS,
	N_("operate on binary package header(s) that would be built from spec"), NULL },
    { "srpm", 0, POPT_ARG_VAL, &source, RPMQV_SPECSRPM,
	N_("operate on source package header generated by spec"), NULL },
    { "queryformat", 0, POPT_ARG_STRING, &queryformat, 0,
	N_("use the following query format"), "QUERYFORMAT" },
    { "qf", 0, (POPT_ARG_STRING | POPT_ARGFLAG_DOC_HIDDEN), &queryformat, 0,
	NULL, NULL },
    POPT_TABLEEND
};

/* the structure describing the options we take and the defaults */
static struct poptOption optionsTable[] = {
    { NULL, '\0', POPT_ARG_INCLUDE_TABLE, specOptsTable, 0,
	N_("Spec options:"), NULL },

    { NULL, '\0', POPT_ARG_INCLUDE_TABLE, rpmcliAllPoptTable, 0,
	N_("Common options for all rpm modes and executables:"), NULL },

   POPT_AUTOALIAS
   POPT_AUTOHELP
   POPT_TABLEEND
};

typedef int (*parsecb)(rpmSpec spec);

#ifdef HAVE_READLINE
static int doShell(rpmSpec spec)
{
    fprintf(stderr, _("RPM version %s macro shell\n"), rpmEVR);
    char *line = NULL;
    while ((line = readline("> ")) != NULL) {
	char *exp = rpmExpand(line, NULL);
	if (*exp)
	    fprintf(stdout, "%s\n", exp);
	free(exp);
	if (*line)
	    add_history(line);
    }
    return 0;
}
#endif

static int dumpSpec(rpmSpec spec)
{
    fprintf(stdout, "%s", rpmSpecGetSection(spec, RPMBUILD_NONE));
    return 0;
}

static int doSpec(poptContext optCon, parsecb cb)
{
    int ec = 0;
    const char * spath;
    char *target = rpmExpand("%{_target}", NULL);
    if (!poptPeekArg(optCon))
	argerror(_("no arguments given for parse"));

    while ((spath = poptGetArg(optCon)) != NULL) {
	rpmSpec spec = rpmSpecParse(spath, (RPMSPEC_ANYARCH|RPMSPEC_FORCE), NULL);
	if (spec) {
	    ec += cb(spec);
	    rpmSpecFree(spec);
	} else {
	    ec++;
	}
	rpmFreeMacros(NULL);
	rpmReadConfigFiles(rpmcliRcfile, target);
    }
    free(target);
    return ec;
}

int main(int argc, char *argv[])
{
    rpmts ts = NULL;
    QVA_t qva = &rpmQVKArgs;

    poptContext optCon;
    int ec = 0;

    xsetprogname(argv[0]); /* Portability call -- see system.h */

    optCon = rpmcliInit(argc, argv, optionsTable);

    if (rpmcliPipeOutput && initPipe())
	exit(EXIT_FAILURE);
	
    ts = rpmtsCreate();
    switch (mode) {

    case MODE_QUERY:
	if (!poptPeekArg(optCon))
	    argerror(_("no arguments given for query"));

	qva->qva_queryFormat = queryformat;
	qva->qva_source = source;
	qva->qva_specQuery = rpmspecQuery;
	ec = rpmcliQuery(ts, qva, (ARGV_const_t) poptGetArgs(optCon));
	break;

    case MODE_PARSE: {
	ec = doSpec(optCon, dumpSpec);
	break;
    }

#ifdef HAVE_READLINE
    case MODE_SHELL:
	if (poptPeekArg(optCon))
	    ec = doSpec(optCon, doShell);
	else
	    ec = doShell(NULL);
	break;
#endif

    case MODE_UNKNOWN:
	if (poptPeekArg(optCon) != NULL || argc <= 1 || rpmIsVerbose()) {
	    printUsage(optCon, stderr, 0);
	    ec = argc;
	}
	break;
    }

    rpmtsFree(ts);
    if (finishPipe())
	ec = EXIT_FAILURE;

    free(qva->qva_queryFormat);

    rpmcliFini(optCon);

    return RETVAL(ec);
}
